src/main/java/net/minecraftforge/client/event/RenderTooltipEvent.java

```diff
@@ -106,39 +133,120 @@ public int getY()
+     */
+    @Cancelable
+    public static class GatherComponents extends Event
+    {
+        private final ItemStack stack;
+        private final int screenWidth;
+        private final int screenHeight;
+        private final List<Either<FormattedText, TooltipComponent>> tooltipElements;
+        private int maxWidth;
+
+        public GatherComponents(ItemStack stack, int screenWidth, int screenHeight, List<Either<FormattedText, TooltipComponent>> tooltipElements, int maxWidth)
+        {
+            this.stack = stack;
+            this.screenWidth = screenWidth;
+            this.screenHeight = screenHeight;
+            this.tooltipElements = tooltipElements;
+            this.maxWidth = maxWidth;
+        }
+
+        /**
+         * @return the ItemStack whose tooltip is being rendered or an empty stack if this tooltip is not for a stack
+         */
+        public ItemStack getStack()
+        {
+            return stack;
+        }
+
+        /**
+         * @return the width of the screen
+         */
+        public int getScreenWidth()
+        {
+            return screenWidth;
+        }
+
+        /**
+         * @return the height of the screen
+         */
+        public int getScreenHeight()
+        {
+            return screenHeight;
+        }
+
+        /**
+         * The elements to be rendered. These can be either formatted text or custom tooltip components.
+         * This list is modifiable.
+         */
+        public List<Either<FormattedText, TooltipComponent>> getTooltipElements()
+        {
+            return tooltipElements;
+        }
+
+        /**
+         * @return the current maximum width for the text components of the tooltip (or -1 for no maximum width)
+         */
+        public int getMaxWidth()
+        {
+            return maxWidth;
+        }
+
+        /**
+         * Set the maximum width for the text components of the tooltip (or -1 for no maximum width)
+         */
+        public void setMaxWidth(int maxWidth)
+        {
+            this.maxWidth = maxWidth;
+        }
+    }
+
     /**
      * This event is fired before any tooltip calculations are done. It provides setters for all aspects of the tooltip, so the final render can be modified.
      * <p>
      * This event is {@link Cancelable}.
      */
-    @net.minecraftforge.eventbus.api.Cancelable
+    @Cancelable
     public static class Pre extends RenderTooltipEvent
     {
         private int screenWidth;
         private int screenHeight;
         private int maxWidth;
 
+        @Deprecated(forRemoval = true) // TODO: remove in 1.18
         public Pre(@Nonnull ItemStack stack, @Nonnull List<? extends FormattedText> lines, PoseStack matrixStack, int x, int y, int screenWidth, int screenHeight, int maxWidth, @Nonnull Font fr)
         {
-            super(stack, lines, matrixStack, x, y, fr);
+            this(stack, matrixStack, x, y, screenWidth, screenHeight, fr, List.of());
```

Comment:
While it is a bit more of a mess, I feel like this should more or less be a copy of the new Pre constructor instead of calling this, so that it can then pass `lines` to the now deprecated `RenderTooltipEvent` super constructor, in case any mods are currently listening to the events and any other mods are still using/decide to use the GuiUtils methods for some reason.